File: py/ventora_billing/src/ventora_billing/webhooks.py
from __future__ import annotations

from typing import cast

import stripe

SUBSCRIPTION_EVENT_TYPES = {
    "customer.subscription.created",
    "customer.subscription.updated",
    "customer.subscription.deleted",
    "customer.subscription.trial_will_end",
}

INVOICE_EVENT_TYPES = {
    "invoice.payment_succeeded",
    "invoice.payment_failed",
}

CHECKOUT_EVENT_TYPES = {
    "checkout.session.completed",
    "checkout.session.expired",
}


def verify_webhook_signature(
    payload: str | bytes,
    signature: str,
    secret: str,
) -> stripe.Event:
    """Verify Stripe webhook signature and return the event. Raises StripeError on failure."""
    return cast(stripe.Event, stripe.Webhook.construct_event(payload, signature, secret))  # type: ignore[no-untyped-call]


def is_subscription_event(event: stripe.Event) -> bool:
    return event.type in SUBSCRIPTION_EVENT_TYPES


def is_invoice_event(event: stripe.Event) -> bool:
    return event.type in INVOICE_EVENT_TYPES


def is_checkout_event(event: stripe.Event) -> bool:
    return event.type in CHECKOUT_EVENT_TYPES


def get_subscription_from_event(event: stripe.Event) -> stripe.Subscription | None:
    if not is_subscription_event(event):
        return None
    obj = event.data.object
    if isinstance(obj, stripe.Subscription):
        return obj
    return None

File: packages/billing/src/webhooks.ts
import type Stripe from "stripe";

export type WebhookEventType =
  | "customer.subscription.created"
  | "customer.subscription.updated"
  | "customer.subscription.deleted"
  | "customer.subscription.trial_will_end"
  | "invoice.payment_succeeded"
  | "invoice.payment_failed"
  | "checkout.session.completed"
  | "checkout.session.expired";

export function verifyWebhookSignature(
  stripe: Stripe,
  body: string,
  signature: string,
  secret: string,
): Stripe.Event {
  return stripe.webhooks.constructEvent(body, signature, secret);
}

const SUBSCRIPTION_EVENT_TYPES = new Set<string>([
  "customer.subscription.created",
  "customer.subscription.updated",
  "customer.subscription.deleted",
  "customer.subscription.trial_will_end",
]);

const INVOICE_EVENT_TYPES = new Set<string>([
  "invoice.payment_succeeded",
  "invoice.payment_failed",
]);

const CHECKOUT_EVENT_TYPES = new Set<string>([
  "checkout.session.completed",
  "checkout.session.expired",
]);

export function isSubscriptionEvent(
  event: Stripe.Event,
): event is Stripe.Event & { data: { object: Stripe.Subscription } } {
  return SUBSCRIPTION_EVENT_TYPES.has(event.type);
}

export function isInvoiceEvent(
  event: Stripe.Event,
): event is Stripe.Event & { data: { object: Stripe.Invoice } } {
  return INVOICE_EVENT_TYPES.has(event.type);
}

export function isCheckoutEvent(
  event: Stripe.Event,
): event is Stripe.Event & { data: { object: Stripe.Checkout.Session } } {
  return CHECKOUT_EVENT_TYPES.has(event.type);
}

export function getSubscriptionFromEvent(event: Stripe.Event): Stripe.Subscription | null {
  if (!isSubscriptionEvent(event)) return null;
  return event.data.object as Stripe.Subscription;
}
